### ISSUE TITLE
redundant code emits redundant instructions

### ISSUE THREAD

**Issue by MkazemAkhgary** [ispc/ispc#2802]
I was testing a few things and noticed this. the `_x` variable is not used here so it shouldn't be compiled.

```c++
uint8 test(uniform uint8<TARGET_WIDTH> x)
{
    uint8 _x;
    unmasked
    {
        _x = x[programIndex]+1;
    }
    return  x[programIndex]+1;
}
```
```asm
test___unT_3C_8_3E_:                    # @test___unT_3C_8_3E_
        vmovmskps       eax, ymm1
        cmp     eax, 255
        je      .LBB1_2
        not     al
        test    al, -127
.LBB1_2:
        vpcmpeqd        xmm1, xmm1, xmm1
        vpsubb  xmm0, xmm0, xmm1
        vzeroupper
        ret
```
---
by removing the redundant part:
```c++
uint8 test(uniform uint8<TARGET_WIDTH> x)
{
    return  x[programIndex]+1;
}
```
```asm
test___unT_3C_8_3E_:                    # @test___unT_3C_8_3E_
        vpcmpeqd        xmm1, xmm1, xmm1
        vpsubb  xmm0, xmm0, xmm1
        ret 
```

I'm not sure what those extra instructions do, looks redundant to me.

**Comment by nurmukhametov** [comment]
Yes, indeed. I think that `unmasked` region doesn't matter here. The following code also produces similar redundancy
```ispc
uint8 test(uniform uint8<TARGET_WIDTH> x)
{
    uint8 y = x[0];
    return x[programIndex]+1;
}
```
```asm
test___unT_3C_8_3E_:                    # @test___unT_3C_8_3E_
        vmovmskps       eax, ymm1
        cmp     eax, 255
        je      .LBB1_2
        not     al
        test    al, -127
.LBB1_2:
        vpcmpeqd        xmm1, xmm1, xmm1
        vpsubb  xmm0, xmm0, xmm1
        vzeroupper
        ret
```

**Comment by dbabokin** [comment]
I would experiment with `SimplifyCFG` and DCE to see if they are able to solve the problem and if they trigger in the right place in the opt pipeline.

**Comment by nurmukhametov** [comment]
> Yes, indeed. I think that `unmasked` region doesn't matter here. The following code also produces similar redundancy
> 
> ```
> uint8 test(uniform uint8<TARGET_WIDTH> x)
> {
>     uint8 y = x[0];
>     return x[programIndex]+1;
> }
> ```
> 
> ```assembly
> test___unT_3C_8_3E_:                    # @test___unT_3C_8_3E_
>         vmovmskps       eax, ymm1
>         cmp     eax, 255
>         je      .LBB1_2
>         not     al
>         test    al, -127
> .LBB1_2:
>         vpcmpeqd        xmm1, xmm1, xmm1
>         vpsubb  xmm0, xmm0, xmm1
>         vzeroupper
>         ret
> ```

For this example with `avx2` targets, we have leftovers of control flow that was generated for `y`/`x[0]` varying variables. ISPC effectively generates two bodies for this function (see [here](https://github.com/ispc/ispc/blob/b697c651a25fb38cc8329a25ef1a80c2de7adebe/src/func.cpp#L471)):
1. The first one is in `some_on` basic block that performs computation for any `mask` value.
2. The second one is in `all_on` basic block that is specialized with `all_on` mask value.

Note that `x[programIndex]+1` is not masked, so the compiler tries to remove all extra operations and related control flow but it doesn't manage to hoist common instructions from successors in `simplifyCFG`. It is a quite limited optimization, see [here](https://github.com/llvm/llvm-project/blob/5997ebdb4feae22e83ec23cc9ee14e3eb7a39e10/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L1591).

I also doubt that duplicating the function body with a specialized path for `all_on` mask is worth doing. The drawbacks are obvious, but the benefits are unclear to me. Testing does not show that this is strictly positive.
